codebase/src/algorithms.ts

```diff
@@ -0,0 +1,56 @@
+import type { AdjacencyList, } from "./graph/types";
+
+/**
+ * 
+ * @param graph in the form of an adjacency list
+ * @param startNode is the label of the node from which the function starts the traversal
+ * @returns a set consisting of the node labels in the order they were visited
+ */
+export function bfsWithTrace(graph: AdjacencyList, source: string): number[] {
+  const startNode = Number(source)
+  const visited = new Set<number>();  // Track visited nodes
+  const queue: number[] = [startNode];  // Queue for BFS traversal
+
+  while (queue.length > 0) {
+    const curr = queue.shift()!;  // Get the first node from the queue
```

Comment:
If I remove the non-null assertion it gives me errors when trying to check if the visited array includes curr because curr could potentially be undefined. It doesn't make sense to me because if that line is being executed, curr cannot be empty due to the while loop checking if the queue was empty.